File: src/test/java/rental_calculator_should.java
import io.vavr.collection.Seq;
import io.vavr.collection.Vector;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class rental_calculator_should {
    private final Seq<Rental> rentals = Vector.of(
            Rental.builder()
                    .amount(1089.90)
                    .date(LocalDate.of(2020, 10, 9))
                    .label("Le Refuge des Loups (LA BRESSE)")
                    .build(),
            Rental.builder()
                    .amount(1276.45)
                    .date(LocalDate.of(2020, 10, 12))
                    .label("Au pied de la Tour (NOUILLORC)")
                    .build(),
            Rental.builder()
                    .amount(670.89)
                    .date(LocalDate.of(2020, 10, 24))
                    .label("Le moulin du bonheur (GLANDAGE)")
                    .build());

    @Test
    public void throws_an_illegal_state_exception_when_no_rentals() {
        assertThrows(IllegalStateException.class, () -> RentalCalculator.calculateRentalAmount.apply(Vector.empty()));
    }

    @Test
    public void calculate_rentals() {
        var amount = RentalCalculator.calculateRentalAmount.apply(rentals);
        assertEquals(3037.24, amount, 0.01);
    }

    @Test
    public void format_statement() {
        var statement = RentalCalculator.formatStatement.apply(rentals);

        assertEquals("2020-10-09 : Le Refuge des Loups (LA BRESSE) | 1089.900000 \n" +
                "2020-10-12 : Au pied de la Tour (NOUILLORC) | 1276.450000 \n" +
                "2020-10-24 : Le moulin du bonheur (GLANDAGE) | 670.890000 \n" +
                "Total amount | 3037.240000", statement);
    }
}

File: scala/src/test/scala/rental_calculator_should.scala
package com.ythirion

import RentalCalculator.{calculateRental, formatStatement}

import org.scalatest.flatspec.AnyFlatSpec

import java.time.LocalDate
import scala.util.Try

class rental_calculator_should extends AnyFlatSpec {
  val rentals = List(
    new Rental(LocalDate.of(2020, 10, 9), "Le Refuge des Loups (LA BRESSE)", 1089.90),
    new Rental(LocalDate.of(2020, 10, 12), "Au pied de la Tour (NOUILLORC)", 1276.45),
    new Rental(LocalDate.of(2020, 10, 24), "Le moulin du bonheur (GLANDAGE)", 670.89),
  )

  "Calculate rentals on a list" should "return the total amount" in {
    val result = calculateRental(rentals)

    assert(result.isSuccess)
    assert(3037.24 == BigDecimal(result.get).setScale(2, BigDecimal.RoundingMode.HALF_UP))
  }

  "Format statement on a list" should "produce a formatted statement" in {
    val statement = formatStatement(rentals)

    assert(statement.isSuccess)
    assert(statement.get ==
      """2020-10-09 : Le Refuge des Loups (LA BRESSE) | 1089.90
        |2020-10-12 : Au pied de la Tour (NOUILLORC) | 1276.45
        |2020-10-24 : Le moulin du bonheur (GLANDAGE) | 670.89
        |Total amount | 3037.24""".stripMargin)
  }

  it should "produce IllegalStateException when rentals is empty on calculation" in {
    assertExceptionCase(calculateRental)
  }

  it should "produce IllegalStateException when rentals is empty on statement formatting" in {
    assertExceptionCase(formatStatement)
  }

  private def assertExceptionCase(f: List[Rental] => Try[Any]): Unit = {
    val result = f(Nil)

    assert(result.isFailure)
    assert(result.failed.get.isInstanceOf[IllegalStateException])
  }
}